# Ejercicios - Clase 6

Ejercicios de los bloques de introducción a python. Numpy

**Versión**: 1.0

**Fecha**: 21/10/2023

En este notebook vamos a trabajar con dos conjuntos de datos (datasets), que han sido creado artificialmente para los ejercicios. Por ahora no debéis saber qué es Pandas, pero lo utilizaremos para leer el fichero en formato CSV, visualizarlo y luego pasarlo a Numpy, librería que sí conocemos. A fin de cuentas, leer un fichero CSV es lo mismo que gestionar una matriz de datos.

El primer dataset define las ventas de los productos de una empresa. Incluye la fecha de venta, el ID del producto, el número de productos vendidos y el precio del producto vendido.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/alfonso-santos/Intro-Python-Finanzas-2023/main/Ejercicios/datos/ventas_generadas.csv', index_col='Unnamed: 0')
df

fecha  producto  productos_vendidos  precio_producto
0      2006-01-18        15                   4            13.53
1      2023-05-31         2                   4             3.57
2      2007-08-01        13                   2             1.78
3      2013-11-17         8                   3             2.11
4      2010-12-01        18                   3            26.59
...           ...       ...                 ...              ...
99995  2003-12-23        17                   2            23.49
99996  2008-04-01        15                   1            13.53
99997  2004-06-09         8                   4             2.11
99998  2017-10-09        13                   2             1.78
99999  2012-06-27        10                   1             3.35

[100000 rows x 4 columns]

In [3]:
datos = df.to_numpy()
datos.shape

(100000, 4)

## Ejercicio 1

Obtén el total de ventas del producto con ID 16 en el año 2008.

In [4]:
id_producto = 16

# TO-DO Primero, define los índices que quieres obtener
# 1. Guarda en la variable indices_producto los índices del id_producto 16
indices_producto = (datos[:, 1] == id_producto)
# 2. Guarda en la variable indices_anno los índices del año 2008
indices_anno = (datos[:, 0].astype("datetime64[Y]") == np.datetime64("2008"))
# 3. Combina los índices entre booleanos.
indices = indices_producto * indices_anno


# Visualizamos los datos con pandas
df[indices]

fecha  producto  productos_vendidos  precio_producto
66     2008-11-28        16                   1             3.75
426    2008-10-23        16                   3             3.75
486    2008-08-30        16                   4             3.75
3234   2008-05-06        16                   2             3.75
3726   2008-11-19        16                   4             3.75
...           ...       ...                 ...              ...
98751  2008-12-12        16                   3             3.75
98925  2008-01-15        16                   4             3.75
99051  2008-10-08        16                   4             3.75
99075  2008-05-30        16                   4             3.75
99501  2008-02-07        16                   4             3.75

[209 rows x 4 columns]

In [5]:
# A partir de aquí, datos_ej1 contiene tu matriz con las filas que quieres procesar

datos_ej1 = datos[indices]
print(datos_ej1.shape)

# TO-DO termina el resto del programa calculando el precio total de ventas
ventas = datos_ej1[:, 2]
ventas_total = ventas.sum()

print("El total de ventas del producto con ID 16 en 2008 es {}".format(ventas_total))

(209, 4)
El total de ventas del producto con ID 16 en 2008 es 542


Respuesta esperada:

- El total de ventas del producto con ID 16 en 2008 es 542

## Ejercicio 2

Define una función que haga todo el proceso anterior de calcular el total de ventas de un producto determinado en un año específico. Completa la siguiente función:

In [6]:
def obtiene_total_ventas(datos, id_producto, anno):
    # TO-DO completa la función
    indices_producto = (datos[:, 1] == id_producto)
    indices_anno = (datos[:, 0].astype("datetime64[Y]") == np.datetime64(str(anno)))
    indices = indices_producto * indices_anno
    
    datos = datos[indices]
    ventas = datos[:, 2]
    return ventas.sum()

Imprime el ID del producto que más se ha vendido en cada año.

In [7]:
# Para que el código funcione, necesitas definir estas 2 variables:
#   anno: El año que se está procesando en el bucle
#   best_producto: El producto más vendido en el año correspondiente

# TO-DO Para cada año, tienes que buscar el mayor número de ventas de cada producto
for anno in np.unique(datos[:, 0].astype("datetime64[Y]")):
    best_ventas = 0
    best_producto = -1
    for id_producto in np.unique(datos[:, 1]):
        ventas = obtiene_total_ventas(datos, id_producto, anno)
        if best_ventas < ventas:
            best_ventas = ventas
            best_producto = id_producto
    print("En el año {} el producto con más ventas es {}".format(anno, best_producto))

En el año 2000 el producto con más ventas es 14
En el año 2001 el producto con más ventas es 7
En el año 2002 el producto con más ventas es 9
En el año 2003 el producto con más ventas es 9
En el año 2004 el producto con más ventas es 8
En el año 2005 el producto con más ventas es 10
En el año 2006 el producto con más ventas es 6
En el año 2007 el producto con más ventas es 17
En el año 2008 el producto con más ventas es 8
En el año 2009 el producto con más ventas es 18
En el año 2010 el producto con más ventas es 19
En el año 2011 el producto con más ventas es 17
En el año 2012 el producto con más ventas es 15
En el año 2013 el producto con más ventas es 13
En el año 2014 el producto con más ventas es 15
En el año 2015 el producto con más ventas es 18
En el año 2016 el producto con más ventas es 12
En el año 2017 el producto con más ventas es 10
En el año 2018 el producto con más ventas es 6
En el año 2019 el producto con más ventas es 17
En el año 2020 el producto con más ventas es 19

Respuesta esperada:

```
En el año 2000 el producto con más ventas es 14
En el año 2001 el producto con más ventas es 7
En el año 2002 el producto con más ventas es 9
En el año 2003 el producto con más ventas es 9
En el año 2004 el producto con más ventas es 8
En el año 2005 el producto con más ventas es 10
En el año 2006 el producto con más ventas es 6
En el año 2007 el producto con más ventas es 17
En el año 2008 el producto con más ventas es 8
En el año 2009 el producto con más ventas es 18
En el año 2010 el producto con más ventas es 19
En el año 2011 el producto con más ventas es 17
En el año 2012 el producto con más ventas es 15
En el año 2013 el producto con más ventas es 13
En el año 2014 el producto con más ventas es 15
En el año 2015 el producto con más ventas es 18
En el año 2016 el producto con más ventas es 12
En el año 2017 el producto con más ventas es 10
En el año 2018 el producto con más ventas es 6
En el año 2019 el producto con más ventas es 17
En el año 2020 el producto con más ventas es 19
En el año 2021 el producto con más ventas es 7
En el año 2022 el producto con más ventas es 1
En el año 2023 el producto con más ventas es 18
```

## Ejercicio 3

Además del producto con más ventas, ahora tienes que calcular cuánto se ha ganado cada año con la venta del producto más vendido.

In [8]:
# Para que el código funcione, necesitas definir estas 3 variables:
#   anno: El año que se está procesando en el bucle
#   best_producto: El producto más vendido en el año correspondiente
#   ganancia: El valor de vender ese producto N veces


# TO-DO Construye un diccionario con el que puedas guardar el precio de cada producto
precio_producto = {}
for id_producto in np.unique(datos[:, 1]):
    indices_producto = (datos[:, 1] == id_producto)
    precio_producto[id_producto] = np.unique(datos[indices_producto, 3])[0]

# TO-DO Para cada año, tienes que buscar el mayor número de ventas de cada producto
#       Además, ahora tendrás que calcular el precio ganado por el mejor
for anno in np.unique(datos[:, 0].astype("datetime64[Y]")):
    best_ventas = 0
    best_producto = -1
    for id_producto in np.unique(datos[:, 1]):
        ventas = obtiene_total_ventas(datos, id_producto, anno)
        if best_ventas < ventas:
            best_ventas = ventas
            best_producto = id_producto
    ganancia = precio_producto[best_producto] * best_ventas
    print("En el año {} el producto con más ventas es {}. Se han ganado {:.2f}€".format(anno, best_producto, ganancia))

En el año 2000 el producto con más ventas es 14. Se han ganado 1567.98€
En el año 2001 el producto con más ventas es 7. Se han ganado 15370.88€
En el año 2002 el producto con más ventas es 9. Se han ganado 12753.00€
En el año 2003 el producto con más ventas es 9. Se han ganado 12687.60€
En el año 2004 el producto con más ventas es 8. Se han ganado 1297.65€
En el año 2005 el producto con más ventas es 10. Se han ganado 2003.30€
En el año 2006 el producto con más ventas es 6. Se han ganado 9329.92€
En el año 2007 el producto con más ventas es 17. Se han ganado 13906.08€
En el año 2008 el producto con más ventas es 8. Se han ganado 1314.53€
En el año 2009 el producto con más ventas es 18. Se han ganado 15555.15€
En el año 2010 el producto con más ventas es 19. Se han ganado 4527.09€
En el año 2011 el producto con más ventas es 17. Se han ganado 14986.62€
En el año 2012 el producto con más ventas es 15. Se han ganado 8483.31€
En el año 2013 el producto con más ventas es 13. Se han ganado 1

Respuesta esperada:

```
En el año 2000 el producto con más ventas es 14. Se han ganado 1567.98€
En el año 2001 el producto con más ventas es 7. Se han ganado 15370.88€
En el año 2002 el producto con más ventas es 9. Se han ganado 12753.00€
En el año 2003 el producto con más ventas es 9. Se han ganado 12687.60€
En el año 2004 el producto con más ventas es 8. Se han ganado 1297.65€
En el año 2005 el producto con más ventas es 10. Se han ganado 2003.30€
En el año 2006 el producto con más ventas es 6. Se han ganado 9329.92€
En el año 2007 el producto con más ventas es 17. Se han ganado 13906.08€
En el año 2008 el producto con más ventas es 8. Se han ganado 1314.53€
En el año 2009 el producto con más ventas es 18. Se han ganado 15555.15€
En el año 2010 el producto con más ventas es 19. Se han ganado 4527.09€
En el año 2011 el producto con más ventas es 17. Se han ganado 14986.62€
En el año 2012 el producto con más ventas es 15. Se han ganado 8483.31€
En el año 2013 el producto con más ventas es 13. Se han ganado 1071.56€
En el año 2014 el producto con más ventas es 15. Se han ganado 8090.94€
En el año 2015 el producto con más ventas es 18. Se han ganado 15262.66€
En el año 2016 el producto con más ventas es 12. Se han ganado 306.87€
En el año 2017 el producto con más ventas es 10. Se han ganado 2103.80€
En el año 2018 el producto con más ventas es 6. Se han ganado 9314.16€
En el año 2019 el producto con más ventas es 17. Se han ganado 13083.93€
En el año 2020 el producto con más ventas es 19. Se han ganado 4483.35€
En el año 2021 el producto con más ventas es 7. Se han ganado 15607.76€
En el año 2022 el producto con más ventas es 1. Se han ganado 13028.90€
En el año 2023 el producto con más ventas es 18. Se han ganado 16166.72€
```

## Ejercicio 4

Para este último ejercicio vamos a utilizar un segundo dataset, también creado artificialmente, que guarda los retornos de 10 activos a lo largo de 1000 días.

In [9]:
retornos = np.loadtxt('https://raw.githubusercontent.com/alfonso-santos/Intro-Python-Finanzas-2023/main/Ejercicios/datos/retornos_generados.csv', delimiter=',', skiprows=1)
retornos.shape

(1000, 10)

El objetivo de este ejercicio consiste en definir de forma aleatoria los pesos de los activos de vuestra cartera. Luego, calcular el retorno global de la cartera y su volatilidad.

In [10]:
n_activos = 10

# Fijamos una semilla a un valor fijo para siempre obtener los mismos resultados
np.random.seed(13)

# TO-DO Utiliza np.random.rand para generar los pesos aleatorios de los activos. No te olvides de normalizarlos
pesos = np.random.rand(n_activos)
pesos /= pesos.sum()

print(pesos)

[0.11142631 0.03403402 0.11809956 0.13836895 0.13935067 0.06496852
 0.08726134 0.11111456 0.09192797 0.10344809]


In [11]:
# TO-DO Calcula el promedio de los retornos de cada activo
retornos_medio = retornos.mean(axis=0) # (5)

# TO-DO Calcula el retorno de la cartera.
retorno_cartera = retornos_medio @ pesos.T * 252

print(retorno_cartera)

0.03411248935042933


In [12]:
# TO-DO Utiliza np.cov para calcular la matriz de covarianzas, la necesitas para calcular la volatilidad
covarianzas = np.cov(retornos.T)

# TO-DO Calcula la varianza de tu cartera
varianza = pesos.T @ covarianzas @ pesos

# TO-DO Calcula la volatilidad de tu cartera
volatilidad = np.sqrt(varianza)

print(f"La rentabilidad de la cartera es {retorno_cartera*100:.2f}% y la volatilidad es {volatilidad*np.sqrt(252)*100:.2f}%")

La rentabilidad de la cartera es 3.41% y la volatilidad es 2.90%


Respuesta esperada fijando la semilla a 13:

- La rentabilidad de la cartera es 3.41% y la volatilidad es 2.90%